lib/TPP/GPU/LinalgToGpu.cpp

```diff
@@ -93,6 +94,200 @@ static bool supportsMMACompute(linalg::LinalgOp linalgOp) {
          cType.getElementType().isF16() && m == 16 && n == 16 && k == 16;
 }
 
+// Fuse a consumer using WMMA operations.
+// Returns updated store op or nullptr if the fusion fails.
+static Operation *mmaFusion(linalg::LinalgOp rootOp, linalg::LinalgOp consumer,
+                            gpu::SubgroupMmaStoreMatrixOp rootStoreOp,
+                            ValueRange storeIndices,
+                            PatternRewriter &rewriter) {
+  Location loc = rootOp.getLoc();
+
+  auto rootOutput = rootOp.getDpsInits()[0];
+  auto outputType = rootOutput.getType().cast<ShapedType>();
+  gpu::MMAMatrixType mmaOutputType = gpu::MMAMatrixType::get(
+      outputType.getShape(), outputType.getElementType(), "COp");
+  auto leadingDim = rootStoreOp.getLeadDimension();
+
+  Value zero = rewriter.create<arith::ConstantIndexOp>(loc, 0);
+
+  // Insert fused eltwise ops before the store and later replace the store
+  // with a new result.
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(rootStoreOp);
+
+  Operation *newStore = nullptr;
+  SmallVector<Value> operands;
+  if (structured_match::utils::isTwoDAddOp(consumer, &operands)) {
+    // Get the value to be added - load the tile first.
+    // Must be a buffer of the same type - scalar broadcast is not supported.
+    auto addValue = operands[0] != rootOutput ? operands[0] : operands[1];
+    if (addValue.getType() != rootOutput.getType())
+      return nullptr;
+    // Fuse the add into the matmul body.
+    addValue = rewriter
+                   .create<gpu::SubgroupMmaLoadMatrixOp>(
+                       loc, mmaOutputType, addValue, ValueRange{zero, zero},
+                       leadingDim,
+                       /*transpose=*/UnitAttr())
+                   .getRes();
+    auto eltwiseAttr = isa<FloatType>(outputType.getElementType())
+                           ? gpu::MMAElementwiseOp::ADDF
+                           : gpu::MMAElementwiseOp::ADDI;
+    auto addRes =
+        rewriter
+            .create<gpu::SubgroupMmaElementwiseOp>(
+                loc, mmaOutputType, ValueRange{rootStoreOp.getSrc(), addValue},
+                eltwiseAttr)
+            .getRes();
+    // Store the new result.
+    newStore = rewriter.replaceOpWithNewOp<gpu::SubgroupMmaStoreMatrixOp>(
+        rootStoreOp, addRes, rootStoreOp.getDstMemref(), ValueRange{zero, zero},
+        leadingDim,
+        /*transpose=*/UnitAttr());
+  } else if (structured_match::utils::isTwoDReluOp(consumer, &operands)) {
+    // Must be a floating point type.
+    auto floatType = dyn_cast<FloatType>(outputType.getElementType());
```

Comment:
why is this not checked for the add? Can we simply constraint the matcher to match only float ops?

Reply:
Constrained to floats only in general.